stubs/Flask-SQLAlchemy/flask_sqlalchemy/__init__.pyi

```diff
@@ -16,37 +18,37 @@ class SignallingSession(Session):
 
 def get_debug_queries(): ...
 
-class Pagination:
-    query: Any
-    page: Any
-    per_page: Any
-    total: Any
+_T = TypeVar("_T")
+
+class BaseQuery(Query[_T]):
+    def get_or_404(self, ident, description: Any | None = ...): ...
+    def first_or_404(self, description: Any | None = ...): ...
+    def paginate(
+        self, page: Any | None = ..., per_page: Any | None = ..., error_out: bool = ..., max_per_page: Any | None = ...
+    ) -> Pagination[_T]: ...
+
+class Pagination(Generic[_T]):
+    query: BaseQuery[_T] | None
+    page: int
+    per_page: int
+    total: int | None
     items: Any
-    def __init__(self, query, page, per_page, total, items) -> None: ...
+    def __init__(self, query: BaseQuery[_T] | None, page: int, per_page: int, total: int | None, items) -> None: ...
     @property
-    def pages(self): ...
-    def prev(self, error_out: bool = ...): ...
+    def pages(self) -> int: ...
+    def prev(self: Self, error_out: bool = ...) -> Self: ...
```

Comment:
`BaseQuery.paginate()` is hardcoded to return a `Pagination` object, so `Pagination.prev()` won't necessarily return whatever the type of `self` is.

To illustrate, if we have a subclass of `Pagination`, `Foo`:

```python
class Foo(Pagination): ...
```

Then the type of `self` in the `Foo.prev()` will be `Foo`, but the returned type will _not_ be `Foo` (unless `Pagination.prev()` is overridden in the subclass); it will still be `Pagination`.

```suggestion
    def prev(self, error_out: bool = ...) -> Pagination: ...
```

Reply:
Ah, good point. I hadn't considered the sub-class use case like that. Thanks for explaining that :)

Reply:
No problem :)